internal/controller/backend_security_policy.go

```diff
@@ -48,7 +65,70 @@ func (b backendSecurityPolicyController) Reconcile(ctx context.Context, req ctrl
 		return ctrl.Result{}, err
 	}
 
+	if oidc := getBackendSecurityPolicyAuthOIDC(backendSecurityPolicy.Spec); oidc != nil {
+		var requeue time.Duration
+		requeue = time.Minute
+		region := backendSecurityPolicy.Spec.AWSCredentials.Region
+
+		rotator, err := rotators.NewAWSOIDCRotator(ctx, b.client, b.kube, b.logger, backendSecurityPolicy.Namespace, backendSecurityPolicy.Name, preRotationWindow, region)
+		if err != nil {
+			b.logger.Error(err, "failed to create AWS OIDC rotator")
+		} else if rotator.IsExpired() {
+			bspKey := fmt.Sprintf("%s.%s", backendSecurityPolicy.Name, backendSecurityPolicy.Namespace)
+
+			var validToken *oauth2.Token
+			if tokenResponse, ok := b.oidcTokenCache[bspKey]; !ok || rotators.IsBufferedTimeExpired(preRotationWindow, tokenResponse.Expiry) {
+				oidcProvider := oauth.NewOIDCProvider(oauth.NewClientCredentialsProvider(b.client), oidc)
+				// Valid Token will be nil if fetch token errors.
+
+				timeOutCtx, cancelFunc := context.WithTimeout(ctx, outGoingTimeOut)
+				defer cancelFunc()
+				validToken, err = oidcProvider.FetchToken(timeOutCtx)
+				if err != nil {
+					b.logger.Error(err, "failed to fetch OIDC provider token")
+				} else {
+					b.oidcTokenCache[bspKey] = validToken
+				}
+			} else {
+				validToken = tokenResponse
+			}
+
+			if validToken != nil {
+				b.oidcTokenCache[bspKey] = validToken
+				awsCredentials := backendSecurityPolicy.Spec.AWSCredentials
+
+				// This is to abstract the real STS behavior for testing purpose.
+				if b.StsOP != nil {
+					rotator.SetSTSOperations(b.StsOP)
+				}
+
+				// Set a timeout for rotate.
+				timeOutCtx, cancelFunc2 := context.WithTimeout(ctx, outGoingTimeOut)
+				defer cancelFunc2()
+				rotator.UpdateCtx(timeOutCtx)
+				token := validToken.AccessToken
+				err = rotator.Rotate(awsCredentials.Region, awsCredentials.OIDCExchangeToken.AwsRoleArn, token)
+				if err != nil {
+					b.logger.Error(err, "failed to rotate AWS OIDC exchange token")
+					requeue = time.Minute
+				} else {
+					requeue = time.Until(rotator.GetPreRotationTime())
+				}
+
+			}
+		}
+		// TODO: Investigate how to stop stale events from re-queuing.
```

Comment:
Let's think about the scenarios:
- If the client secret is reset, the previous enqueued client secret will fail to get the token from the OIDC provider. 
- If the AWS IAM roles is updated, the previous enqueued ARN will fail to exchange the AWS credentials. 
These failed events may clog the controller and keeps retrying.

Reply:
Actually Requeue only enqueue the reconciliation key, so we may not need to worry about the stale event.